src/cryptonote_core/pulse.cpp

```diff
@@ -687,319 +983,421 @@ event_loop prepare_for_round(round_context &context, service_nodes::service_node
-    auto our_bitset   = std::bitset<sizeof(validator_bitset) * 8>(context.submit_block_template.validator_bitset);
-    MERROR(log_prefix(context) << "Internal error, unexpected block validator bitset is empty " << block_bitset << ", our bitset was " << our_bitset);
-    return event_loop::keep_running;
+    // Message
+    pulse::message msg = msg_init_from_context(context);
+    msg.type           = pulse::message_type::signed_block;
+    crypto::generate_signature(msg_signature_hash(context, msg), key.pub, key.key, msg.signature);
+    handle_message(quorumnet_state, msg); // Add our own. We receive our own msg for the first time which also triggers us to relay.
   }
 
-  // TODO(doyle): Does this need to be == expected hashes or can it just be > min validator nodes for signing, i.e. 7.
-  if (timed_out || received_hashes == expected_hashes)
+  //
+  // NOTE: Wait
+  //
+  handle_messages_received_early_for(context.transient.signed_block.wait.stage, quorumnet_state);
+  pulse_wait_stage const &stage = context.transient.signed_block.wait.stage;
+
+  auto const &quorum   = context.transient.signed_block.wait.data;
+  bool const timed_out = pulse::clock::now() >= stage.end_time;
+  bool const enough    = stage.msgs_received >= context.transient.wait_for_handshake_bitsets.best_count;
+
+  if (timed_out || enough)
   {
-    if (timed_out && received_hashes != expected_hashes)
+    if (!enforce_validator_participation_and_timeouts(context, stage, timed_out, enough))
       return goto_preparing_for_next_round(context);
 
-    context.state = round_state::wait_for_next_block;
-    auto received_bitset = std::bitset<sizeof(validator_bitset) * 8>(context.submit_block_template.validator_bitset);
-    MINFO(log_prefix(context) << "Received " << received_hashes << " random value hashes from " << received_bitset << (timed_out ? ". We timed out and some hashes are missing" : ""));
-    return event_loop::keep_running;
+    // Select signatures randomly so we don't always just take the first N required signatures.
+    // Then sort just the first N required signatures, so signatures are added
+    // to the block in sorted order, but were chosen randomly.
+    std::array<size_t, service_nodes::PULSE_QUORUM_NUM_VALIDATORS> indices = {};
+    size_t indices_count = 0;
+
+    // Pull out indices where we've received a signature
+    for (size_t index = 0; index < quorum.size(); index++)
+    {
+      bool received = quorum[index].second;
+      if (received) indices[indices_count++] = index;
+    }
+
+    // Shuffle and sort first 'N' PULSE_BLOCK_REQUIRED_SIGNATURES entries
+    assert(indices_count >= service_nodes::PULSE_BLOCK_REQUIRED_SIGNATURES);
+    tools::shuffle_portable(indices.begin(), indices.begin() + indices_count, tools::rng);
+    std::sort(indices.begin(), indices.begin() + service_nodes::PULSE_BLOCK_REQUIRED_SIGNATURES);
```

Comment:
It doesn't look like this needs to be reproducible, so this shuffle+sort could be simplified using std::sample to do a one-pass, order-preserving sample (this might need change indices to be uint16_t as well):
```C++
std::array<uint16_t, service_nodes::PULSE_BLOCK_REQUIRED_SIGNATURES> selected;
std::sample(indices.begin(), indices.begin() + indices_count, selected.begin(), selected.size(), tools::rng);
```

Reply:
And updated.